torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -117,7 +135,28 @@ void PropagateShapeOnNode(Node * node) {
       node->output()->inferTypeFrom(node->t(kvalue));
     } break;
     case kUndefined: {
-      node->output()->setType(nullptr);
+      node->output()->setType(DynamicType::get());
+    } break;
+    case kIf: {
+      auto then_block = node->blocks()[0];
+      auto else_block = node->blocks()[1];
+      PropagateShapeOnBlock(then_block);
+      PropagateShapeOnBlock(else_block);
+      mergeTypes(then_block->outputs(), else_block->outputs(), node->outputs());
+    } break;
+    case kLoop: {
+      auto body_block = node->blocks()[0];
+      // propagate counter type
+      body_block->inputs()[0]->setType(node->inputs()[0]->type());
+      // propagate loop-carried input types to block inputs
+      auto loop_carried_inputs = node->inputs().slice(2); // skip max, cond
+      auto loop_carried_block = body_block->inputs().slice(1); // skip trip
+      for(size_t i = 0; i < loop_carried_inputs.size(); ++i) {
+        loop_carried_block[i]->setType(loop_carried_inputs[i]->type());
+      }
+      PropagateShapeOnBlock(body_block);
+      auto loop_carried_outputs = body_block->outputs().slice(1); // skip cond
+      mergeTypes(loop_carried_inputs, loop_carried_outputs, node->outputs());
```

Comment:
This doesn't look correct to me. If this merger fails, you should also reset all the types inside the body block to Dynamic (since they will run with the outer types once, and if the loop doesn't terminate immediately, they will also see the new ones).

Reply:
Yep, cood catch -- I updated  it to iterate the propagation of types in the body until we reach convergence. This will take at most N cycles where N is the number of loop carried dependencies.